{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/hpcfs/home/ciencias_biologicas/na.portilla10/phallet/Metrics_Results/Kuttervirus', '/hpcfs/home/ciencias_biologicas/na.portilla10/phallet/Metrics_Results/Tequintavirus', '/hpcfs/home/ciencias_biologicas/na.portilla10/phallet/Metrics_Results/Cheoctovirus', '/hpcfs/home/ciencias_biologicas/na.portilla10/phallet/Metrics_Results/Skunavirus', '/hpcfs/home/ciencias_biologicas/na.portilla10/phallet/Metrics_Results/Epseptimavirus']\n",
      "/hpcfs/home/ciencias_biologicas/na.portilla10/phallet/Metrics_Results/Kuttervirus\n",
      "/hpcfs/home/ciencias_biologicas/na.portilla10/phallet/Metrics_Results/Tequintavirus\n",
      "/hpcfs/home/ciencias_biologicas/na.portilla10/phallet/Metrics_Results/Cheoctovirus\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "# Set working directory\n",
    "workdir = os.path.expanduser(\"~/phallet/Metrics_Results\")\n",
    "subdirectories = [os.path.join(workdir, name) for name in os.listdir(workdir) if os.path.isdir(os.path.join(workdir, name))]\n",
    "print(subdirectories)\n",
    "\n",
    "# Read command-line arguments\n",
    "args = sys.argv[1:]\n",
    "mx = \"ani\"\n",
    "kmersx = [12, 11, 10, 9, 8]\n",
    "my = \"mash\"\n",
    "kmersy = [15, 17, 20, 21, 24]\n",
    "\n",
    "# Remove any arguments that are not valid integers\n",
    "args = [arg for arg in args if re.match(r'^\\d+$', arg)]\n",
    "\n",
    "# Cases for metrics selected on Y\n",
    "if my == \"mash\":\n",
    "    tool_my = [\"mash\", \"sourmash\"]\n",
    "elif my == \"ani\":\n",
    "    tool_my = [\"fastani\", \"skani\"]\n",
    "elif my == \"aai\":\n",
    "    tool_my = [\"comparem\"]\n",
    "elif my == \"viridic\":\n",
    "    tool_my = [\"viridic\"]\n",
    "elif my == \"vcontact2\":\n",
    "    tool_my = [\"vcontact2\"]\n",
    "    \n",
    "if mx == \"mash\":\n",
    "    tool_mx = [\"mash\", \"sourmash\"]\n",
    "elif mx == \"ani\":\n",
    "    tool_mx = [\"fastani\", \"skani\"]\n",
    "elif mx == \"aai\":\n",
    "    tool_mx = [\"comparem\"]\n",
    "elif mx == \"viridic\":\n",
    "    tool_mx = [\"viridic\"]\n",
    "elif mx == \"vcontact2\":\n",
    "    tool_mx = [\"vcontact2\"]\n",
    "\n",
    "metrics = [mx, my]\n",
    "\n",
    "ani_data = pd.DataFrame()\n",
    "mash_data = pd.DataFrame()\n",
    "viridic_data = pd.DataFrame()\n",
    "mx_data = pd.DataFrame()\n",
    "my_data = pd.DataFrame()\n",
    "df_subset_mx = pd.DataFrame()\n",
    "df_subset_my = pd.DataFrame()\n",
    "kmer_values_mx = []\n",
    "kmer_values_my = []\n",
    "\n",
    "for m in metrics:\n",
    "    if m == mx:\n",
    "        kmers = kmersx\n",
    "        tool_mx = tool_mx\n",
    "    if m == my:\n",
    "        kmers = kmersy\n",
    "        tool_my = tool_my\n",
    "\n",
    "for genus in subdirectories:\n",
    "    os.chdir(genus)\n",
    "    print(os.getcwd())\n",
    "    genus_name = os.path.basename(genus)\n",
    "    genus_dir = os.path.join(workdir, genus_name)\n",
    "    summaries = glob.glob(os.path.join(genus_dir, \"*metrics*.csv\"))\n",
    "\n",
    "    if mx in metrics:\n",
    "        mx_files = [s for s in summaries if f\"{mx}_metrics_\" in s]\n",
    "        if len(mx_files) > 0:\n",
    "            mx_data = pd.read_csv(mx_files[0])    \n",
    "        else:\n",
    "            print(f\"No {mx} data for {genus_name}\")\n",
    "    if my in metrics:\n",
    "        my_files = [s for s in summaries if f\"{my}_metrics_\" in s]\n",
    "        if len(my_files) > 0:\n",
    "            my_data = pd.read_csv(my_files[0])\n",
    "        else:\n",
    "            print(f\"No {my} data for {genus_name}\")\n",
    "            \n",
    "# Assuming merged_df is your DataFrame\n",
    "    for algorithm_mx in tool_mx:\n",
    "        for algorithm_my in tool_my:\n",
    "        # Subset the data for the current combination\n",
    "            if \"algorithm\" in mx_data.columns:\n",
    "                df_subset_mx = mx_data[mx_data[\"algorithm\"] == algorithm_mx] \n",
    "                kmer_values_mx = sorted(pd.unique(df_subset_mx[f\"kmer_{mx}\"])) \n",
    "\n",
    "            if \"algorithm\" in my_data.columns:\n",
    "                df_subset_my = my_data[my_data[\"algorithm\"] == algorithm_my]\n",
    "                kmer_values_my = sorted(pd.unique(df_subset_my[f\"kmer_{my}\"]))\n",
    "   \n",
    "# Create a scatterplot for each kmer combination\n",
    "            fig, axes = plt.subplots(len(kmer_values_mx), len(kmer_values_my), figsize=(12,8), sharex=True)\n",
    "            axes = axes.reshape(len(kmer_values_mx), len(kmer_values_my))\n",
    "            \n",
    "            kmer_values_mx.sort()\n",
    "            kmer_values_my.sort(reverse=True)\n",
    "# Check if there are any valid combinations of kmer_values_mx and kmer_values_my\n",
    "            fig.text(0.04, 0.04, algorithm_my, va='center', rotation='vertical')\n",
    "\n",
    "            for i, mx_kmer in enumerate(kmer_values_mx):\n",
    "                for j, my_kmer in enumerate(kmer_values_my):\n",
    "        # Subset the data for the current kmer_mx and kmer_my\n",
    "                    df_subset_mx_kmer = df_subset_mx[df_subset_mx[f\"kmer_{mx}\"] == mx_kmer] \n",
    "                    df_subset_my_kmer = df_subset_my[df_subset_my[f\"kmer_{my}\"] == my_kmer]\n",
    "                    # Check if there are any data points for the current kmer combination\n",
    "                    merged_df = pd.merge(df_subset_mx_kmer, df_subset_my_kmer, on=[\"GenomeA\", \"GenomeB\"], how=\"outer\")\n",
    "                    merged_df.sort_values(by=[f\"{mx}_distance\",f\"{my}_distance\"], inplace=True)\n",
    "        # Create a scatterplot for the current kmer_mx and kmer_my\n",
    "                    ax = axes[i,j]\n",
    "                    ax.set_title(f\"{algorithm_mx} kmer: {kmer_values_mx[i]}\",fontsize=6)\n",
    "                    ax.tick_params(axis='both', which='major',labelsize=4) \n",
    "                    ax.scatter(merged_df[f\"{mx}_distance\"],merged_df[f\"{my}_distance\"],alpha=0.5,s=0.5)\n",
    "                    ax.set_ylabel(f\"{algorithm_my} kmer: {kmer_values_my[j]}\", rotation=90, ha='center',fontsize=6)\n",
    "                    ax.yaxis.set_label_coords(-0.2, 0.5) \n",
    "        #Set the order of the subplot\n",
    "            plt.subplots_adjust(left=0.1,right=0.9, top=0.9, bottom=0.1, wspace=0.1, hspace=0.2)\n",
    "                              \n",
    "        # Save the current figure to a PDF file if there are valid combinations\n",
    "            pdf_filename = f\"{genus_name}_{algorithm_mx}_{algorithm_my}.pdf\"\n",
    "            with PdfPages(pdf_filename) as pdf:\n",
    "                pdf.savefig(fig)\n",
    "      "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "jupyter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
